server/src/main/java/org/elasticsearch/index/engine/ReadOnlyEngine.java

```diff
@@ -122,27 +124,40 @@ public ReadOnlyEngine(EngineConfig config, SeqNoStats seqNoStats, TranslogStats
         }
     }
 
-    protected void ensureMaxSeqNoEqualsToGlobalCheckpoint(final SeqNoStats seqNoStats) {
-        // Before 8.0 the global checkpoint is not known and up to date when the engine is created after
-        // peer recovery, so we only check the max seq no / global checkpoint coherency when the global
-        // checkpoint is different from the unassigned sequence number value.
-        // In addition to that we only execute the check if the index the engine belongs to has been
-        // created after the refactoring of the Close Index API and its TransportVerifyShardBeforeCloseAction
-        // that guarantee that all operations have been flushed to Lucene.
-        final Version indexVersionCreated = engineConfig.getIndexSettings().getIndexVersionCreated();
-        if (indexVersionCreated.onOrAfter(Version.V_7_1_0) ||
-            (seqNoStats.getGlobalCheckpoint() != SequenceNumbers.UNASSIGNED_SEQ_NO && indexVersionCreated.onOrAfter(Version.V_6_7_0))) {
-            if (seqNoStats.getMaxSeqNo() != seqNoStats.getGlobalCheckpoint()) {
-                throw new IllegalStateException("Maximum sequence number [" + seqNoStats.getMaxSeqNo()
-                    + "] from last commit does not match global checkpoint [" + seqNoStats.getGlobalCheckpoint() + "]");
+    /**
+     * Ensures that all operations exist in translog are committed into the given index commit. This check is mandatory
+     * so peer recovery of closed indices can skip phase 2 (i.e., not replaying translog operations) without losing data.
+     */
+    private void ensureNoUncommittedOperation(SeqNoStats seqNoStats, SegmentInfos segmentInfos) throws IOException {
+        // we can't enforce this check on an old index - should we prevent this engine as a recovery source?
```

Comment:
what would that mean? Can we restrict this more ie. doesn't it only apply if we used to be a follower engine?

Reply:
Yes, I pushed https://github.com/elastic/elasticsearch/pull/41317/commits/8ba5ca5b70f8da88d6d5a8f29d4d4d7ec276ecd4 to skip this check if max_seq_no equals to the global checkpoint.